library/core/src/fmt/rt.rs

```diff
@@ -65,61 +65,67 @@ pub struct Argument<'a> {
     ty: ArgumentType<'a>,
 }
 
-#[rustc_diagnostic_item = "ArgumentMethods"]
-impl Argument<'_> {
-    #[inline]
-    const fn new<'a, T>(x: &'a T, f: fn(&T, &mut Formatter<'_>) -> Result) -> Argument<'a> {
+macro_rules! argument_new {
+    ($t:ty, $x:expr, $f:expr) => {
         Argument {
             // INVARIANT: this creates an `ArgumentType<'a>` from a `&'a T` and
             // a `fn(&T, ...)`, so the invariant is maintained.
             ty: ArgumentType::Placeholder {
-                value: NonNull::from_ref(x).cast(),
+                value: NonNull::<$t>::from_ref($x).cast(),
                 // SAFETY: function pointers always have the same layout.
-                formatter: unsafe { mem::transmute(f) },
+                formatter: |ptr: NonNull<()>, fmt: &mut Formatter<'_>| {
+                    let func = $f;
+                    // SAFETY: This is the same type as the `value` field.
```

Comment:
it could be made more apparently that `as_ref` is the unsafe operation, rather than `ptr.cast()`

Reply:
tbh, if someone mistakenly thinks `ptr.cast()` is unsafe, then I don't know if we can help them. The library should be written clearly and sometimes that will mean "for a somewhat experienced reader" instead of someone on some mailing list or social media site offering decontextualized commentary.

Reply:
> instead of someone on some mailing list or social media site offering decontextualized commentary.

Was this necessary? I wasn't expecting bullying here.

Reply:
Apologies. I _am_ trying to restrain myself but Peter Zijlstra's bullying was dutifully linked, so I am probably poorly calibrated at the moment.

Reply:
The safety comment needs to explain why creating a `&T` with the specific type `T` is okay here, so the cast is quite integral to the unsafe operation happening here. I think it's reasonable to keep it inside the unsafe block so it's closer to the safety comment in this case.